opaque_keys/edx/django/models.py

```diff
@@ -194,6 +194,12 @@ class LearningContextKeyField(OpaqueKeyField):
     """
     description = "A LearningContextKey object, saved to the DB in the form of a string"
     KEY_CLASS = LearningContextKey
+    # Declare the field types for the django-stubs mypy type hint plugin.
+    # See https://github.com/typeddjango/django-stubs/blob/6a850f6/django-stubs/db/models/fields/__init__.pyi#L503-L511
+    # for examples of how this works for the standard field types in the upstream codebase.
+    # Note that these particular type annotations have no effect at runtime nor on Django itself nor on PyLance.
+    _pyi_private_set_type: LearningContextKey | str | None  # The types that you can set into this field.
+    _pyi_private_get_type: LearningContextKey  # The type that you get when you read from this field
```

Comment:
So I've seen this pattern elsewhere, and I totally believe you're doing the right thing. But I don't understand this part–why is `None` a settable type, but not a gettable type for this field?

Reply:
@ormsbee 

> why is None a settable type, but not a gettable type for this field?

My understanding was that for these type definitions you generally don't specify `None` as either a setter or a getter, and the mypy plugin will automatically add the `| None` union type if you declare a specific field instance with `null=True`. And I noticed that even if you don't declare it with `null=True`, you can still set a `None` value because the field type [converts](https://github.com/openedx/opaque-keys/blob/e0ee525df4395f3795501fdc3a75a7ecdbedcb1f/opaque_keys/edx/django/models.py#L131-L133) that to an empty string.

That said, I think I was not really thinking things through. `OpaqueKeyField`s should never be declared as `null=True`, because they're `CharField`-based, but instead they should use `blank=True` or not to indicate whether they accept empty values. The problem is that `OpaqueKeyField` will convert empty strings to `None` in the getter, which mypy is unaware of without `null=True`. So in the end, I think it's best to just explicitly specify `None` as one of the possible types on both the getter and the setter, which I have now done.

Thanks for the helpful question. I have pushed a commit to add `| None` to the getters.